[PLY] Fix for 450

runette opened the pull request:
fixes #450 
fixes #452 

Updates to the PLY driver :

- Changes the vertices names to lower case : "x", "y", "z"
- Adds the ability to save in binary mode
- changes the default behaviour to save in binary mode to avoid precision issues
- adds the "format" metadata item to maintain round-trip consistency and allow the format to be changed
- fixes a bug in the handling of CRS data

The resulting binary and text files have been tested in CloudCompare and Blender and open succesfully

runette commented:
@vcloarec @PeterPetrik - please review

vcloarec commented on mdal/frmts/mdal_ply.cpp:
```diff
@@ -577,15 +580,44 @@ void MDAL::DriverPly::save( const std::string &fileName, const std::string &mesh
+  libply::Metadata meta;
+
+  meta.emplace( "crs", mesh->crs() );
+  MDAL::Metadata metadata = mesh->metadata;
```
```suggestion
  const MDAL::Metadata &metadata = mesh->metadata;
```

vcloarec commented on mdal/frmts/mdal_ply.cpp:
```diff
@@ -577,15 +580,44 @@ void MDAL::DriverPly::save( const std::string &fileName, const std::string &mesh
+    { "binary_little_endian", libply::File::Format::BINARY_LITTLE_ENDIAN }
+  };
+
+  for ( auto it = metadata.begin(); it != metadata.end(); ++it )
```
```suggestion
  for ( auto it = metadata.cbegin(); it != metadata.cend(); ++it )
```

vcloarec commented on mdal/frmts/mdal_ply.cpp:
```diff
@@ -577,15 +580,44 @@ void MDAL::DriverPly::save( const std::string &fileName, const std::string &mesh
+
+  for ( auto it = metadata.begin(); it != metadata.end(); ++it )
+  {
+    std::pair< std::string, std::string > item = *it;
```
```suggestion
    const std::pair< std::string, std::string > &item = *it;
```

vcloarec commented on mdal/frmts/mdal_ply.cpp:
```diff
@@ -577,15 +580,44 @@ void MDAL::DriverPly::save( const std::string &fileName, const std::string &mesh
+  MDAL::Metadata metadata = mesh->metadata;
+  libply::File::Format format = libply::File::Format::BINARY_LITTLE_ENDIAN;
+
+  const std::unordered_map<std::string, libply::File::Format> FORMAT_MAP =
```
any reasons to be upper case?

runette replied:
Nope. That is a brain_fart! I think it was an unconscious association with the line above :(

vcloarec reviewed (commented):
Nothing particular to say except some `const` . Do you think it is possible to avoid to  decrease the test coverage?

runette commented:
> Nothing particular to say except some `const` . Do you think it is possible to avoid to decrease the test coverage?

I will take a look at the detailed comments. On the general question - I was worried about that as well (even though it is only 0.02%) and did look at it - but the coverage calculation is a bit strange as it is complaining about 4 lines of code that are definitely tested!

I will see what I can do. I think it is because big-endian is not tested. I will try that.

vcloarec commented:
I am ok with the coverage decrease (only -0.02%), but I am concerned about that we are under 90%, even it is not you fault here.

runette commented:
> I am ok with the coverage decrease (only -0.02%), but I am concerned about that we are under 90%, even it is not you fault here.



> I am ok with the coverage decrease (only -0.02%), but I am concerned about that we are under 90%, even it is not you fault here.

I will have a look at it this evening and see if I can find a way to get it above 90

runette commented:
@vcloarec  @PeterPetrik 

It should now be green - I am not sure why it is not.

I did not get coverage to 90% but got it going the right way.

I increased test coverage to include big_endian files (and in the process discovered and fixed two bugs - so I guess testing is good after all :). I also deleted some unreachable code.

I tested the outputs from this driver in all formats with CloudCompare and Blender and it seems to be working

I propose that it is merged
